Supporters version 2.0

joemull opened the pull request:
This is a comprehensive overhaul of the plugin, based on our current needs and priorities as a diamond open access publisher operating as part of a university that provides some billing support, and using several external billing agents with some groups of supporters.

With this upgrade, I've used our latest and best practices for plugin structure--at least as far as I know!--and it has full test coverage with careful use of mocks, setups, and teardowns to mitigate interference with other tests.

The data model includes a [proposed new methodology](https://birkbeckuol.sharepoint.com/:w:/r/sites/ac-ts-arts-olh/Shared%20Documents/General/OLH%20team/Website/Website%20redevelopment%20project%202023/project%20research%20and%20plan/Proposal%20for%20calculating%20OLH%20supporter%20fees.docx?d=w74eb400818054d9ea4161f7b932fe9df&csf=1&web=1&e=Ky8ZZJ) for calculating the factors that go into someone's support fee. In the proposed breakdown, you can manually set how much supporter/consortium size and support level will affect the fee. And for the external factors--currency exchange rate and economic disparity between countries around the world--the plugin now has support for fetching exchange rate and GNI per capita data and using them in the fee calculation. Using GNI per capita data by country is provisional and we can still reconsider it. This prototype just shows how it might work.

## New data model

The data model is rewritten to avoid duplicating data and to provide better machine readability to meet the fee calculation requirement, as well as to be more legible to staff when managing supporters and bands via the admin interface. The migrations nearly completely preserve the data by using three consecutive migrations: a field and model expansion migration, a data migration, and a cleanup and renaming migration. But the data will require some touch-up afterward (such as with billing agents).

- Closes #17 
- Closes #29 

## Simplified and refined views and forms
- Closes #11
- Closes #21
- Closes #26 
- Closes #30
- Closes #31
- Closes #32 
- Closes #34
- Closes #35
- Closes #41 
- Closes #40

## Easier configuration
- Closes #10 
- Closes #27
- Closes #33
- Closes #38
- Closes #42

## Removed unused features for polling
- Closes #15
- Closes #28

## Removed unused features for renewals, referrals, and billing

Because we manage renewals and billing outside the plugin, these features are no longer needed.

- Closes #1 
- Closes #9  
- Closes #13
- Closes #14 
- Closes #16

ajrbyers commented on views.py:
```diff
@@ -1,728 +1,200 @@
-    return render(request, 'consortial_billing/process_renewal.html', context)
-
-
+# Temporarily adding this security decorator
```
Should this be dropped from the PR?

ajrbyers commented on forms.py:
```diff
@@ -1,113 +1,78 @@
+        self.fields['fee'].required = False
+        self.fields['fee'].disabled = True
+
+    def save(self, commit=True):
```
Can you add a quick explaination of the save methods behavior? It could be considered a little confusing that the save() method might discard the input and return a different object.

ajrbyers commented on templatetags/supporters.py:
Remove empty file?

ajrbyers commented on templates/consortial_billing/manager.html:
```diff
@@ -0,0 +1,301 @@
+                    </p>
+                    <div class="stacked button-group">
+                        <a
+                            href="/admin/consortial_billing/supporter/"
```
It is possible to reverse admin urls (https://docs.djangoproject.com/en/dev/ref/contrib/admin/#reversing-admin-urls)

ajrbyers reviewed (approved):
As discussed, this is quite a difficult one ot reivew. I've added a few comments but I thnk we can merge this and take it forward.

mauromsl commented on management/commands/fetch_world_bank_data.py:
```diff
@@ -0,0 +1,35 @@
+                    )
+                )
+            else:
+                logger.info(
+                    self.style.WARNING(
+                        f'Could not get {year} data'
+                    )
```
If we intend to run this programatically, we should log an ERROR here

mauromsl commented on utils.py:
```diff
@@ -0,0 +1,257 @@
+    return data
+
+
+def generate_new_demo_data():
+    return [
+        make_table_of_higher_supporters_by_country_and_level(),
+        make_table_of_standard_supporters_by_country_and_size()
+    ]
+
+
+def update_demo_band_data():
+    data_json = json.dumps(generate_new_demo_data(), separators=(',', ':'))
+    filename = os.path.join(plugin_settings.SHORT_NAME, DEMO_DATA_FILENAME)
+    return save_media_file(filename, data_json)
+
+
+def get_saved_demo_band_data():
+    filename = os.path.join(plugin_settings.SHORT_NAME, DEMO_DATA_FILENAME)
+    file = cms_models.MediaFile.objects.get(label=filename)
+    with file.file.open('r') as file_ref:
+        return json.loads(file_ref.read())
```
Once you move to using Decimal values, python's JSON library does not have codecs for `decima.Decimal`, you will need to either pass a codec or manually transform the values as you encode/decode

mauromsl commented on tests/test_models.py:
```diff
@@ -0,0 +1,292 @@
+            return_value=(0.85, ''),
+        ) as latest_multiplier:
+            fee, _ = self.band_other_two.calculate_fee()
+            expected_fee = round(
+                5000        # base
+                # * 0.6       # size does not matter for higher level
+                * 0.85      # Patched GNI
+                * 0.85,     # Patched exchange rate
+                -1
+            )
```
Please use decimals, not floats

mauromsl commented on models.py:
```diff
@@ -1,268 +1,539 @@
+        help_text='How size is determined, e.g. 0-4999 FTE staff '
+                  'for an institution'
+    )
+    multiplier = models.FloatField(
```
Please use `DecimalField` https://docs.djangoproject.com/en/5.0/ref/models/fields/#decimalfield

mauromsl reviewed (changes requested):
I've added a few comments on calculations and/or numeric data types regarding the use of floating point arithmetic and `decimal.Decimal`.:

https://docs.python.org/3/library/decimal.html

In short, floating point arithmetic sacrifices precision for performance, which can be useful in scenarios like graphics rendering but is not acceptable here. There is also extensive use of fractions, for which Decimal has  precision problems of its own and it is best to use `Fraction`: https://docs.python.org/3/library/fractions.html

Finally, since you have data coming in and out of the Python runtime (I can see data input/output as JSON and also as part of settings stored as strings), you'll need to make sure any and every input is converted to consistent datatypes. I'd argue even using type annotations here, so that static analysis can assist with wrong datatypes.

joemull commented:
> I've added a few comments on calculations and/or numeric data types regarding the use of floating point arithmetic and `decimal.Decimal`.:
> 
> https://docs.python.org/3/library/decimal.html
> 
> In short, floating point arithmetic sacrifices precision for performance, which can be useful in scenarios like graphics rendering but is not acceptable here. There is also extensive use of fractions, for which Decimal has precision problems of its own and it is best to use `Fraction`: https://docs.python.org/3/library/fractions.html
> 
> Finally, since you have data coming in and out of the Python runtime (I can see data input/output as JSON and also as part of settings stored as strings), you'll need to make sure any and every input is converted to consistent datatypes. I'd argue even using type annotations here, so that static analysis can assist with wrong datatypes.

This was really helpful, thanks. I've now incorporated decimals thoroughly into the app.

mauromsl reviewed (approved):
Nice, good work!
